server/schedule/placement/region_rule_cache.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2021 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package placement
+
+import (
+	"sync"
+
+	"github.com/tikv/pd/server/core"
+)
+
+// RegionRuleFitCacheManager stores each region's RegionFit Result and involving variables
+// only when the RegionFit result is satisfied with its rules
+// RegionRuleFitCacheManager caches RegionFit result for each region only when:
+// 1. region have no down peers
+// 2. RegionFit is satisfied
+// RegionRuleFitCacheManager will invalid the cache for the region only when:
+// 1. region peer topology is changed
+// 2. region have down peers
+// 3. region leader is changed
+// 4. any involved rule is changed
+// 5. stores topology is changed
+// 6. any store label is changed
+// 7. any store state is changed
+type RegionRuleFitCacheManager struct {
+	mu     sync.RWMutex
+	caches map[uint64]*RegionRuleFitCache
```

Comment:
Will it cost too much memory to cache all region's fit result?

Reply:
I think it won't. each `RegionRuleFitCache` stored pointer for `rule`/`peer` which are already in memory. Thus, it won't cost too much memory here.